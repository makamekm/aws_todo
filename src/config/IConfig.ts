import * as ORM from "typeorm";
import {
    LogLevel,
} from "../log";

export interface ITypeORMConfigScope {
    [name: string]: ORM.ConnectionOptions;
}

export interface IAdditionalInformation {
    [name: string]: (IAdditionalInformation | string | number | boolean);
}

export interface IRedisConfigScope {
    [name: string]: {
        port: number;
        host: string;
        password?: string;
        scope?: string;
    };
}

export interface IConfigScope {
    [name: string]: IConfig;
}

export interface IConfig {
    // Log to console with level: [debug, info, warn, error] ["info", "warn", "debug", "error"]
    logLevel?: LogLevel;
    // additional log messages to console [string[]]
    additionalLogMessages?: string[];
    // ORM via typeorm [scope: typeorm object]
    db?: ITypeORMConfigScope;
    // client config in state will be available in browser store [json object]
    publicConfig?: any;
}
