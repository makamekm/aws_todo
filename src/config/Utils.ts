function replaceStandardEnv(str: string) {
    let result: RegExpExecArray;
    // tslint:disable-next-line
    while (result = /\$\{(\w+)\}/gi.exec(str)) {
        str = str.replace("${" + result[1] + "}", process.env[result[1]]);
    }
    return str;
}

function replaceDefaultEnv(str: string) {
    let result: RegExpExecArray;
    // tslint:disable-next-line
    while (result = /\$\{(\w+):(\w+)\}/gi.exec(str)) {
        str = str.replace("${" + result[1] + ":" + result[2] + "}", process.env[result[1]] || result[2]);
    }
    return str;
}

export function replaceEnvVars(str: string) {
    let processedStr = str;
    processedStr = replaceStandardEnv(processedStr);
    processedStr = replaceDefaultEnv(processedStr);
    return processedStr;
}
