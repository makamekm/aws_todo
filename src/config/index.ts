export * from "./IConfig";
export * from "./DefaultConfig";
export * from "./Operations";
