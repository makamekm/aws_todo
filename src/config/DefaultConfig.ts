import { IConfigScope } from "./IConfig";

const config: IConfigScope = {
  default: {
    logLevel: "info",
    db: {
      default: {
        database: "todo",
        host: "localhost",
        password: "",
        port: 5432,
        type: "postgres",
        username: "root",
        synchronize: true,
      },
    },
    publicConfig: {
      graphqlEndpoint: "http://localhost:3000/graphql",
    },
  },
};

export default config;
