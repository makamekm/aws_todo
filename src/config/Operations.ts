import * as fs from "fs";
import * as path from "path";
import defaultConfig from "./DefaultConfig";
import { IConfig, IConfigScope } from "./IConfig";
import { replaceEnvVars } from "./Utils";

const CONFIG_PATH = path.resolve(process.env.CONFIG_PATH || "./config.json");
const CONFIG_NAME = process.env.CONFIG_SCOPE || "default";

let config: IConfigScope;
applyDefaultConfig();

export function applyDefaultConfig() {
    config = cloneConfig(defaultConfig);
}

function parseConfig(configString: string): IConfigScope {
    return JSON.parse(replaceEnvVars(configString));
}

function cloneConfig(configState: IConfigScope): IConfigScope {
    return parseConfig(JSON.stringify(configState));
}

function stringifyConfig(configState: IConfigScope): string {
    return JSON.stringify(configState, null, 2);
}

export function getConfig(): IConfig {
    return config[CONFIG_NAME];
}

export function setConfig(configState: IConfigScope) {
    config = {
        ...config,
        ...cloneConfig(configState),
    };
}

export function readConfig(_fs = fs) {
    if (_fs.existsSync(CONFIG_PATH)) {
        config = {
            ...config,
            ...parseConfig(_fs.readFileSync(CONFIG_PATH, "utf8")),
        };
    } else {
        throw new Error(`The config does not exist using a path: ${CONFIG_PATH}`);
    }
}

export const saveConfig = (_fs = fs) => {
    if (_fs.existsSync(CONFIG_PATH)) {
        _fs.writeFileSync(CONFIG_PATH, stringifyConfig(config), {
            encoding: "utf8",
        });
    } else {
        _fs.writeFileSync(CONFIG_PATH, stringifyConfig(config), {
            encoding: "utf8",
            flag: "wx",
        });
    }
};
