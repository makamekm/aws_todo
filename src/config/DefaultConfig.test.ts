import config from "./DefaultConfig";

describe("Config", () => {
    describe("DefaultConfig", () => {
        it("should match snapshot", () => {
            expect(config).toMatchSnapshot();
        });
    });
});
