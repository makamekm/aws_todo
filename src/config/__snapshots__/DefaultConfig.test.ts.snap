// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Config DefaultConfig should match snapshot 1`] = `
Object {
  "default": Object {
    "db": Object {
      "default": Object {
        "database": "todo",
        "host": "localhost",
        "password": "",
        "port": 5432,
        "synchronize": true,
        "type": "postgres",
        "username": "root",
      },
    },
    "logLevel": "info",
    "publicConfig": Object {
      "graphqlEndpoint": "http://localhost:3000/graphql",
    },
  },
}
`;
